sgv2-service-common/src/main/java/io/stargate/sgv2/common/schema/SchemaCache.java

```diff
@@ -0,0 +1,178 @@
+              }
+
+              // Removed
+              for (String keyspaceName : new HashSet<>(keyspaces.keySet())) {
+                if (!newKeyspaces.containsKey(keyspaceName)) {
+                  deleteKeyspace(keyspaceName);
+                }
+              }
+
+              // If any changes were recorded, they will be replayed now. We don't know which ones
+              // we missed, but reapplying them is idempotent, at worst it's a bit of duplicated
+              // work.
+              observer.markReady();
+
+              return null;
+            });
+  }
+
+  private void updateKeyspace(CqlKeyspaceDescribe newKeyspace, boolean isInit) {
+    String keyspaceName = newKeyspace.getCqlKeyspace().getName();
+    CqlKeyspaceDescribe oldKeyspace = keyspaces.put(keyspaceName, newKeyspace);
+    if (isInit) {
+      assert oldKeyspace == null;
+      // don't send notifications
+    } else if (oldKeyspace == null) {
+      LOG.debug("Added keyspace {}", keyspaceName);
+      notifyListeners(l -> l.onCreateKeyspace(newKeyspace));
+    } else {
+      LOG.debug("Updated keyspace {}", keyspaceName);
+      notifyListeners(l -> l.onUpdateKeyspace(oldKeyspace, newKeyspace));
+    }
+  }
+
+  private void deleteKeyspace(String keyspaceName) {
+    CqlKeyspaceDescribe oldKeyspace = keyspaces.remove(keyspaceName);
+    if (oldKeyspace != null) {
+      LOG.debug("Removed keyspace {}", keyspaceName);
+      notifyListeners(l -> l.onDropKeyspace(oldKeyspace));
+    }
+  }
+
+  private void notifyListeners(Consumer<SchemaListener> action) {
+    for (SchemaListener listener : listeners) {
+      try {
+        action.accept(listener);
+      } catch (Throwable t) {
+        LOG.error("Unexpected error while invoking " + listener, t);
+      }
+    }
+  }
+
+  private class Observer extends InitReplayingStreamObserver<SchemaNotification> {
+
+    @Override
+    protected void onNextReady(SchemaNotification notification) {
+      SchemaChange change = notification.getChange();
+      if (notification.hasKeyspace()) {
+        updateKeyspace(notification.getKeyspace(), false);
+      } else {
+        assert change.getChangeType() == Type.DROPPED && change.getTarget() == Target.KEYSPACE;
+        deleteKeyspace(change.getKeyspace());
+      }
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      LOG.error("Notification stream failed, trying to reconnect", t);
+      tryReconnect();
+    }
+
+    @Override
+    public void onCompleted() {
+      LOG.error("Notification stream completed unexpectedly, trying to reconnect");
+      tryReconnect();
+    }
+
+    private void tryReconnect() {
+      refreshKeyspaces(false)
+          .exceptionally(
+              t -> {
+                LOG.error(
+                    "Error while trying to reconnect to notification stream, "
+                        + "the schema cache will stop updating",
+                    t);
+                return null;
```

Comment:
I'm not sure what else to do here. Maybe some kind of periodic retry, but we don't have access to an executor.